### ISSUE TITLE
php-mcrypt has been removed in php 7.2 which will be used by ubuntu 18.04

### ISSUE THREAD

**Issue by bradymiller** [openemr/openemr#1451]
info here:
http://php.net/manual/en/migration71.deprecated.php

Need to do analysis to see where this is used in OpenEMR and if a fix is needed.

**Comment by bradymiller** [comment]
Looks like the `Enable Audit Log Encryption` in Globals->Logging requires this, which is of course a critical MU2 requirement ...

**Comment by bradymiller** [comment]
Looks like php-mcrypt is getting migrated from core to PHP PECL. So will be better if support this encryption via OpenSSL, if possible. More research needed.

**Comment by bradymiller** [comment]
https://stackoverflow.com/questions/43482840/php-convert-mcrypt-encrypt-to-openssl-encrypt-soap-header

**Comment by sunsetsystems** [comment]
Why is PHP always breaking stuff in new releases? :)

**Comment by bradymiller** [comment]
yep, they like to keep us all on our toes. I suppose it's better than still using PHP3, though :)

**Comment by bradymiller** [comment]
Man, it is just keeps getting a little worse: https://stackoverflow.com/questions/42265194/how-to-decrypt-a-string-with-openssl-which-was-previously-encrypted-with-mcrypt

**Comment by bradymiller** [comment]
If above link is true, we have an interesting dilemma. All log comments that were encrypted in 5.0.0 will not be viewable or able to decrypt in 5.0.1 after upgrade. This is not the end of the world since one could always bring the log tables to a 5.0.0 instance for auditing if needed; but it is a real pain. In a way it does make the changes easier since do not need to deal with the prior log data on the upgrade I suppose. Will keep researching this.

**Comment by bradymiller** [comment]
Code proposal here:
https://github.com/openemr/openemr/pull/1455

After get the code dealt with, then need to remove mcrypt from dependencies in ubuntu packages and dockers (in dockers(and double check after install ubuntu packages), need to ensure php-openssl is installed).

**Comment by sunsetsystems** [comment]
Looks like patient documents might also be encrypted with mcrypt? See controllers/C_Document.class.php.
Might need to find a substitute library.

**Comment by bradymiller** [comment]
Hi @sunsetsystems ,

Note that is an odd feature and I think does the opposite of what you think. As I recall, it doesn't encrypt on storage, but encrypts the file when the user downloads it(and then decrypts when you upload it).

Also, we may get lucky, and that cipher may be supported by openssh.

Also plan to think of making the the audit log stuff a bit more standardized (use a different $iv for each entry, which seems to be like a salt), and possibly store a random key that is created rather than use the same key for everybody.

-brady

**Comment by bradymiller** [comment]
Hi @sunsetsystems ,

Found a better solution that does not require mcrypt (or even openssl) and is using a library that OpenEMR already has. Check out the PR. Initial testing looks good. Still need to test to ensure that stuff that was previosly encrypted will decrypt correctly.

-brady

**Comment by sunsetsystems** [comment]
@bradymiller Tested with mcrypt unavailable? I saw some mcrypt stuff in phpseclib but I guess it handles the case where it's not present. Assuming yes that's excellent!

**Comment by sunsetsystems** [comment]
Looks like we're OK if phpseclib is reasonably up to date: https://github.com/phpseclib/phpseclib/issues/1229

**Comment by bradymiller** [comment]
hi @sunsetsystems ,

Just checked in a different strategy that upgrades the encryption scheme (to use injection vectors and a secret, unique key), which will also be backward compatible to show the prior encrypted log messages via mcrypt (if the user has the mycrpt module installed). Note I didn't try to make backward compatible with the document encryption since it's encrypted on client side and can be dealt with by client, if need (I would pose that this feature is used bu 0% of users anyways).

I still got some testing to go.

Thoughts?
-brady

**Comment by bradymiller** [comment]
@sjpadgett ,

2 questions regarding this:

1. I have noted use of the aes256Encrypt() and aes256Decrypt()  in the new patient portal. Note I upgraded these functions (https://github.com/openemr/openemr/pull/1455) since mcrypt is going away and also made them more secure. It should work for what you are doing right now; but if users encrypted stuff previously, then it won't be able to decrypt them without a little development work (check out what I did to support the "old" encrypted stuff in the log in the PR).

2. In the patient portal I am seeing the `is_msg_encrypted` column in the onsite_mail, but for the life of me, I can't figure out where the encryption/decryption is happening (a codebase search for `is_msg_encrypted`) shows nothing. Am missing something?

thanks,
-brady
